### PR TITLE
Adding support for pointer events

### PR DESCRIPTION
**Pull request by themoonrat**
And for devices that don't support pointer events, it normalises mouse / touch events to give the same data that a pointer event would

Re-submit of https://github.com/pixijs/pixi.js/pull/2476

### CODE REVIEW

**Comment by kaansoral** [comment]
I just discovered a critical issue that prevents all mouse events for a small subset of PC users

One player I worked with had touch-action: none on his canvas

The issue wasn't on 4.0.2, but it was on the dev 4.0.3 build I was using, I wonder whether if this is the cause

**Comment by staff0rd** [comment]
So this causes #3167, and more specifically does not support any backwards compatibility, in that `mousedown`, `mouseup` etc are no longer called on devices that support pointer events.  Is this expected?

**Comment by themoonrat** [comment]
Not expected, no :/ It appears that browsers that support the pointet events will not send out mouse events if the pointer ones are listened to. I shall investigate further.

**Comment by GoodBoyDigital** [comment]
Hey chaps! This one has turned into a pretty big issue for one of our projects thats about to launch. IE11 what a pain huh?

@themoonrat is this something that ya think is a straight forward issue to address? Ideally I would love to have this sorted for a minor release on Wednesday. If its a bit of a tricky one I suggest that we revert this commit so that the issue is resolved for users. We can then take our time to understand how and why IE11 is being a pain. 

Let me know your thoughts!

Cheers

**Comment by themoonrat** [comment]
@GoodBoyDigital I'll take a look tonight, and let you know if I get anywhere / can get anywhere / cry

**Comment by GoodBoyDigital** [comment]
good luck man! Il be there with the hanky if needed 👍

**Comment by kaansoral** [comment]
I would appreciate a PostMortem too, this issue hurt me when it mattered the most, while my game was receiving initial attention, I've lost a lot of potential players: https://github.com/pixijs/pixi.js/issues/3158 (Not blaming anyone, but just saddens me)

Also on https://github.com/pixijs/pixi.js/issues/3167 There's some advice related to using "pointerdown" to replace "mousedown"+"touchdown"

So basically, assuming the issue is fixed or patched, do we use "pointerdown" along with "mousedown"+"touchdown"? (I assume so)
Anyway, to sum up, a usage scenario+example would be nice

**Comment by themoonrat** [comment]
Hi @kaansoral 
First of all, very sorry about the bug.... I switched my game over to just use pointer events so never noticed the issue :/ 

Regarding usage: I personally just use pointer events now. For browsers that support them natively, those are sent through. For browsers that don't support them natively, mouse&touch events get converted into a pointer events with the required properties from the pointer spec added in.

However, the regular mouse and touch events are still sent through (now, after this PR with the fix).

So my advice is EITHER 
listen to mouse AND touch events
OR
listen to JUST pointer events.

**Comment by kaansoral** [comment]
Ah, thanks

I started refactoring now, basically replacing "mousedown","touchstart" with "pointerdown"

However, "rightdown" worried me, when the issue happened, both left and right clicks weren't working, and as far as I checked, there is no "rightdown" replacement in the pointer event-set?

(I also read the Code, checked for a documentation, but came up empty)

**Comment by themoonrat** [comment]
This is the page you are looking for: https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events
Search for 'Determining button states'

To get to this info on a PIXI event for pointerdown, you'd look at
event.data.originalEvent.button or event.data.originalEvent.buttons

**Comment by kaansoral** [comment]
Thanks for the info, is it verified that "rightdown" works as before on devices that have pointer events enabled? (I mean testing-wise, I can't test it myself, as I don't have such a device/system)

**Comment by themoonrat** [comment]
'rightdown' should still be sent on devices that support pointer events. I'll verify this on some devices in a few hours time when I have access to them and the little test page I made.

(ps. the latest chrome versions now support pointer events, so it's not just an IE & Edge thing anymore)

**Comment by kaansoral** [comment]
I deployed 4.1.1 - People started reporting the inability to click again, so I reversed to 4.0.2 :(

**Comment by themoonrat** [comment]
4.1.1 didn't include anything other than a blank .npmignore file from 4.1.0 as described in the release notes https://github.com/pixijs/pixi.js/releases/tag/v4.1.1
I believe a 4.2.0 release is due soonish which would contain this pointer event fix

**Comment by kaansoral** [comment]
Ah, apologies then, I assumed the minor release @GoodBoyDigital mentioned was 4.1.1

Looking forward to 4.2.0 then :)

**Comment by englercj** [comment]
We use semantic versioning (http://semver.org), where the version is defined as `Major.Minor.Patch`.
